// @mintlify/validation/dist/openapi/convertSecurity.js
import { InvalidSchemaError } from './convertOpenApi.js';
export const convertSecurity = ({ securityRequirements, securitySchemes, }) => {
    if (securityRequirements === undefined || securityRequirements.length === 0) {
        return [];
    }
    if (securitySchemes === undefined) {
        throw new InvalidSchemaError(['#', 'components'], 'securitySchemes not defined');
    }
    // TODO(ronan): make this work for camel-case as well
    return securityRequirements.map((security) => {
        const title = Object.keys(security)
            .map((securityName) => securityName.replace(/[_-]/g, ' '))
            .join(' & ');
        const parameterSections = {
            title,
            query: {},
            header: {},
            cookie: {},
        };
        Object.keys(security).forEach((securityName) => {
            const securityScheme = securitySchemes === null || securitySchemes === void 0 ? void 0 : securitySchemes[securityName];
            if (securityScheme === undefined) {
                throw new InvalidSchemaError(['#', 'components', 'securitySchemes'], `security scheme not defined: '${securityName}'`);
            }
            addSecurityParameters({
                securityName,
                securityScheme,
                parameterSections,
            });
        });
        return parameterSections;
    });
};
export const addSecurityParameters = ({ securityName, securityScheme, parameterSections, }) => {
    var _a, _b;
    switch (securityScheme.type) {
        case 'apiKey':
            if (!['header', 'query', 'cookie'].includes(securityScheme.in)) {
                throw new InvalidSchemaError(['#', 'components', 'securitySchemes', securityName], `invalid security scheme location provided: '${securityScheme.in}'`);
            }
            const paramGroup = securityScheme.in;
            parameterSections[paramGroup][securityScheme.name] = {
                description: securityScheme.description,
                required: true,
                deprecated: false,
                schema: [
                    {
                        type: 'string',
                        required: true,
                        deprecated: false,
                    },
                ],
            };
            return;
        case 'http':
            if (securityScheme.scheme === 'basic') {
                parameterSections.header['Authentication'] = {
                    description: (_a = securityScheme.description) !== null && _a !== void 0 ? _a : 'Basic authentication header of the form `Basic <encoded-value>`, where `<encoded-value>` is the base64-encoded string `username:password`.',
                    required: true,
                    schema: 'basic',
                    deprecated: false,
                };
            }
            else if (securityScheme.scheme === 'bearer') {
                parameterSections.header['Authorization'] = {
                    description: (_b = securityScheme.description) !== null && _b !== void 0 ? _b : 'Bearer authentication header of the form `Bearer <token>`, where `<token>` is your auth token.',
                    required: true,
                    schema: 'bearer',
                    deprecated: false,
                };
            }
            else {
                throw new InvalidSchemaError(['#', 'components', 'securitySchemes', securityName], `encountered unknown HTTP security scheme: '${securityScheme.scheme}'`);
            }
            return;
        case 'oauth2':
        case 'openIdConnect':
            return;
    }
};
