// @mintlify/validation/dist/openapi/convertOpenApi.js
import { convertParameters } from './convertParameters.js';
import { convertSchema } from './convertSchema.js';
import { convertSecurity } from './convertSecurity.js';
import { convertServers } from './convertServers.js';
export const generateMessage = (path, messages = []) => {
    const pathString = path
        .map((component) => component.replace('\\', '\\\\').replace('/', '\\/'))
        .join('/');
    return [pathString, ...messages].join('\n');
};
export class InvalidSchemaError extends Error {
    constructor(path, ...messages) {
        super(generateMessage(path, messages));
        this.name = 'InvalidSchemaError';
        Object.setPrototypeOf(this, InvalidSchemaError.prototype);
    }
}
export class ImpossibleSchemaError extends Error {
    constructor(path, ...messages) {
        super(generateMessage(path, messages));
        this.name = 'ImpossibleSchemaError';
        Object.setPrototypeOf(this, ImpossibleSchemaError.prototype);
    }
}
export class ConversionError extends Error {
    constructor(path, ...messages) {
        super(generateMessage(path, messages));
        this.name = 'ConversionError';
        Object.setPrototypeOf(this, ConversionError.prototype);
    }
}
export const convertBody = (path, body) => {
    if (body === undefined) {
        return {};
    }
    const newEntries = Object.entries(body.content).map(([contentType, mediaObject]) => {
        return [
            contentType,
            convertSchema([...path, contentType, 'schema'], mediaObject.schema, body.required),
        ];
    });
    return Object.fromEntries(newEntries);
};
export const convertResponses = (path, responses) => {
    const newEntries = Object.entries(responses).map(([statusCode, response]) => [
        statusCode,
        convertResponse([...path, statusCode], response),
    ]);
    return Object.fromEntries(newEntries);
};
export const convertResponse = (path, response) => {
    if (response.content === undefined) {
        return {};
    }
    const newEntries = Object.entries(response.content).map(([contentType, mediaType]) => [
        contentType,
        convertSchema([...path, 'content', contentType, 'schema'], mediaType.schema),
    ]);
    return Object.fromEntries(newEntries);
};
export const convertOpenAPIV3_1ToEndpoint = (spec, path, method) => {
    var _a, _b, _c, _d, _e;
    const paths = spec.paths;
    if (paths === undefined) {
        throw new InvalidSchemaError(['#'], 'paths not defined');
    }
    const pathObject = paths[path];
    if (pathObject === undefined) {
        throw new InvalidSchemaError(['#', 'paths'], `path not defined: ${path}`);
    }
    const operationObject = pathObject[method];
    if (operationObject === undefined) {
        throw new InvalidSchemaError(['#', 'paths', path], `operation does not exist: ${method}`);
    }
    const securityRequirements = (_a = operationObject.security) !== null && _a !== void 0 ? _a : spec.security;
    const securitySchemes = (_b = spec.components) === null || _b === void 0 ? void 0 : _b.securitySchemes;
    const security = convertSecurity({
        securityRequirements,
        securitySchemes,
    });
    const pathParameters = pathObject.parameters;
    const operationParameters = operationObject.parameters;
    const parameters = convertParameters({
        path: ['#', 'paths', path],
        method,
        pathParameters: pathParameters,
        operationParameters: operationParameters,
    });
    const servers = convertServers({
        servers: (_d = (_c = operationObject.servers) !== null && _c !== void 0 ? _c : pathObject.servers) !== null && _d !== void 0 ? _d : spec.servers,
    });
    const description = (_e = operationObject.description) !== null && _e !== void 0 ? _e : pathObject === null || pathObject === void 0 ? void 0 : pathObject.description;
    const requestBody = operationObject.requestBody;
    const body = convertBody(['#', 'paths', path, method, 'requestBody'], requestBody);
    const deprecated = !!operationObject.deprecated;
    const response = convertResponses(['#', 'paths', path, method, 'responses'], operationObject.responses);
    return {
        description,
        path,
        method,
        servers,
        request: {
            security,
            parameters,
            body,
        },
        response,
        deprecated,
    };
};
